clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,748 @@
+                                   GlobalVariable::InternalLinkage, Arr, Name);
+    Var->setUnnamedAddr(GlobalValue::UnnamedAddr::Global);
+
+    SmallVector<char, 32> NameBuf;
+    auto SectionName = Section.toStringRef(NameBuf);
+    if (!SectionName.empty())
+      Var->setSection(SectionName);
+    return Var;
+  }
+
+  // Adds given buffer as a global variable into the module and returns a pair
+  // of pointers that point to the beginning and the end of the variable.
+  std::pair<Constant *, Constant *>
+  addArrayToModule(ArrayRef<char> Buf, const Twine &Name,
+                   const Twine &Section = "") {
+    auto *Var = addGlobalArrayVariable(Name, Buf, Section);
+    auto *ImageB = ConstantExpr::getGetElementPtr(Var->getValueType(), Var,
+                                                  getSizetConstPair(0, 0));
+    auto *ImageE = ConstantExpr::getGetElementPtr(
+        Var->getValueType(), Var, getSizetConstPair(0, Buf.size()));
+    return std::make_pair(ImageB, ImageE);
+  }
+
+  // Adds given data buffer as constant byte array and returns a constant
+  // pointer to it. The pointer type does not carry size information.
+  Constant *addRawDataToModule(ArrayRef<char> Data, const Twine &Name) {
+    auto *Var = addGlobalArrayVariable(Name, Data);
+    auto *DataPtr = ConstantExpr::getGetElementPtr(Var->getValueType(), Var,
+                                                   getSizetConstPair(0, 0));
+    return DataPtr;
+  }
+
+  // Creates all necessary data objects for the given image and returns a pair
+  // of pointers that point to the beginning and end of the global variable that
+  // contains the image data.
+  std::pair<Constant *, Constant *>
+  addDeviceImageToModule(ArrayRef<char> Buf, const Twine &Name,
+                         StringRef TargetTriple) {
+    // Create global variable for the image data.
+    // TODO: recheck TargetTriple.empty()
+    return addArrayToModule(
+        Buf, Name,
+        TargetTriple.empty() ? ""
+                             : "__CLANG_OFFLOAD_BUNDLE__sycl-" + TargetTriple);
+  }
+
+  // Creates a global variable of const char* type and creates an
+  // initializer that initializes it with given string (with added null
+  // terminator). Returns a link-time constant pointer (constant expr) to that
+  // variable.
+  Constant *addStringToModule(StringRef Str, const Twine &Name) {
+    auto *Arr = ConstantDataArray::getString(C, Str);
+    auto *Var = new GlobalVariable(M, Arr->getType(), /*isConstant*/ true,
+                                   GlobalVariable::InternalLinkage, Arr, Name);
+    Var->setUnnamedAddr(GlobalValue::UnnamedAddr::Global);
+    auto *Zero = ConstantInt::get(getSizeTTy(), 0);
+    Constant *ZeroZero[] = {Zero, Zero};
+    return ConstantExpr::getGetElementPtr(Var->getValueType(), Var, ZeroZero);
+  }
+
+  // Creates a global variable of array of structs and initializes
+  // it with the given values in @ArrayData.
```

Comment:
Note: LLVM uses [Doxygen](https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments) for generating documentation and this and other comments do not follow the right syntax